docs/design/cpsubsystem/02-group-availability-listener.md

```diff
@@ -0,0 +1,145 @@
+# CP Subsystem Group Availability Listener
+
+|ℹ️ Since: 4.1| 
+|-------------|
+
+## Background
+
+CP Subsystem requires manual user intervention in some cases, especially CP members crash or become unreachable. 
+But there's no programmatic way of monitoring CP members. Users should track Hazelcast logs or look at Management Center 
+to see any problems in CP membership healthiness. 
+
+Our goal in this work to add a new listener to the CP Subsystem to get notified when availability of a CP group decreases. 
+This also covers majority loss of a CP group.
+
+
+## API Design
+
+Two new methods will be added to the CPSubsystem interface, to register and deregister listeners.
+
+```java
+    /**
+     * Registers a new CPGroupAvailabilityListener to listen CP group availability changes.
+     *
+     * @param listener group availability listener
+     * @return id of the listener registration
+     * @since 4.1
+     */
+    UUID addGroupAvailabilityListener(CPGroupAvailabilityListener listener);
+
+    /**
+     * Removes CPGroupAvailabilityListener registration.
+     *
+     * @param id of the registration
+     * @return true if listener registration is removed, false otherwise
+     * @since 4.1
+     */
+    boolean removeGroupAvailabilityListener(UUID id);
+```                        
+
+A new `CPGroupAvailabilityListener` interface and a new `CPGroupAvailabilityEvent` interface will be added to the 
+CP Subsystem module under `com.hazelcast.cp.event` package.
+
+```java
+/**
+ * CPGroupAvailabilityListener is notified when availability
+ * of a CP group decreases or it loses the majority completely.
+ * @since 4.1
+ */
+public interface CPGroupAvailabilityListener extends EventListener {
+
+    /**
+     * Called when a CP group's availability decreases,
+     * but still has the majority of members available.
+     *
+     * @param event CP group availability event
+     */
+    void availabilityDecreased(CPGroupAvailabilityEvent event);
+
+    /**
+     * Called when a CP group has lost its majority.
+     *
+     * @param event CP group availability event
+     */
+    void majorityLost(CPGroupAvailabilityEvent event);
+}
+```       
+
+```java 
+/**
+ * CPGroupAvailabilityEvent is published when a CP group's
+ * availability is decreased or it has lost the majority completely.
```

Comment:
I'm wondering whether we could (or it makes sense) fire an event whenever the CP group majority is back again. Kind of like "split brain healed" kind of thing.

Reply:
I thought about it too. But it's not very straightforward. Because in this case, we are monitoring the CP group from outside, I mean outside of the Raft consensus. That's why we track unavailability of a member based on Hazelcast's failure detector. If a member is removed from cluster, it's clear that availability is decreased. But a member joining to cluster does not mean it's still part of the CP group and can interact in consensus. It might be a different member started on the same address, it might have lost the CP data. We cannot observe that from outside of the group.

Reply:
If it's requested too, later we can add them. It needs more work.

Reply:
Ok, agreed. Since it's not so trivial, let's not invest more time in it unless requested.